NPE in Redis mapping response to `OptimisticLockingTransactionResult`

### Describe the bug

In https://github.com/quarkusio/quarkus/blob/67ba9c06afa7d0adcf8f3d2b4bba8dc3488b8ba1/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/TransactionHolder.java#L37 mapping fails for `null` `responsePart`. `null` value is correctly checked in https://github.com/quarkusio/quarkus/blob/67ba9c06afa7d0adcf8f3d2b4bba8dc3488b8ba1/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/TransactionHolder.java#L21

### Expected behavior

No NPE for `null` `responsePart`

### Actual behavior

```
Caused by: java.lang.NullPointerException: Cannot invoke "io.vertx.mutiny.redis.client.Response.type()" because "responsePart" is null
	at io.quarkus.redis.runtime.datasource.TransactionHolder.toOptimisticLockingResult(TransactionHolder.java:42)
	at io.quarkus.redis.runtime.datasource.ReactiveRedisDataSourceImpl.toTransactionResult(ReactiveRedisDataSourceImpl.java:222)
	at io.quarkus.redis.runtime.datasource.BlockingRedisDataSourceImpl.withTransaction(BlockingRedisDataSourceImpl.java:181)
```

### How to Reproduce?

https://github.com/luca-bassoricci/issue_52478_reproducer

### Output of `uname -a` or `ver`

windows 11

### Output of `java -version`

java 21

### Quarkus version or git rev

3.31.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/BlockingRedisDataSourceImpl.java
+++ b/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/BlockingRedisDataSourceImpl.java
@@ -1,6 +1,9 @@
 package io.quarkus.redis.runtime.datasource;
 
 import static io.quarkus.redis.runtime.datasource.ReactiveRedisDataSourceImpl.toTransactionResult;
+import static io.quarkus.redis.runtime.datasource.Validation.notNullOrEmpty;
+import static io.smallrye.mutiny.helpers.ParameterValidation.doesNotContainNull;
+import static io.smallrye.mutiny.helpers.ParameterValidation.nonNull;
 
 import java.time.Duration;
 import java.util.function.BiConsumer;
@@ -63,7 +66,8 @@ public class BlockingRedisDataSourceImpl implements RedisDataSource {
         this(new ReactiveRedisDataSourceImpl(vertx, redis, connection), timeout);
     }
 
-    public TransactionResult withTransaction(Consumer<TransactionalRedisDataSource> ds) {
+    public TransactionResult withTransaction(Consumer<TransactionalRedisDataSource> tx) {
+        nonNull(tx, "tx");
         RedisConnection connection = reactive.redis.connect().await().atMost(timeout);
         ReactiveRedisDataSourceImpl dataSource = new ReactiveRedisDataSourceImpl(reactive.getVertx(), reactive.redis,
                 connection);
@@ -74,7 +78,7 @@ public class BlockingRedisDataSourceImpl implements RedisDataSource {
         try {
             connection.send(Request.cmd(Command.MULTI)).await().atMost(timeout);
             try {
-                ds.accept(source);
+                tx.accept(source);
             } catch (Exception e) {
                 if (!source.discarded()) {
                     try {
@@ -97,7 +101,10 @@ public class BlockingRedisDataSourceImpl implements RedisDataSource {
     }
 
     @Override
-    public TransactionResult withTransaction(Consumer<TransactionalRedisDataSource> ds, String... watchedKeys) {
+    public TransactionResult withTransaction(Consumer<TransactionalRedisDataSource> tx, String... watchedKeys) {
+        nonNull(tx, "tx");
+        notNullOrEmpty(watchedKeys, "watchedKeys");
+        doesNotContainNull(watchedKeys, "watchedKeys");
         RedisConnection connection = reactive.redis.connect().await().atMost(timeout);
         ReactiveRedisDataSourceImpl dataSource = new ReactiveRedisDataSourceImpl(reactive.getVertx(), reactive.redis,
                 connection);
@@ -113,7 +120,7 @@ public class BlockingRedisDataSourceImpl implements RedisDataSource {
             connection.send(cmd).await().atMost(timeout);
             connection.send(Request.cmd(Command.MULTI)).await().atMost(timeout);
             try {
-                ds.accept(source);
+                tx.accept(source);
             } catch (Exception e) {
                 if (!source.discarded()) {
                     try {
@@ -138,6 +145,10 @@ public class BlockingRedisDataSourceImpl implements RedisDataSource {
     @Override
     public <I> OptimisticLockingTransactionResult<I> withTransaction(Function<RedisDataSource, I> preTx,
             BiConsumer<I, TransactionalRedisDataSource> tx, String... watchedKeys) {
+        nonNull(preTx, "preTx");
+        nonNull(tx, "tx");
+        notNullOrEmpty(watchedKeys, "watchedKeys");
+        doesNotContainNull(watchedKeys, "watchedKeys");
         RedisConnection connection = reactive.redis.connect().await().atMost(timeout);
         ReactiveRedisDataSourceImpl dataSource = new ReactiveRedisDataSourceImpl(reactive.getVertx(), reactive.redis,
                 connection);

--- a/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactiveRedisDataSourceImpl.java
+++ b/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactiveRedisDataSourceImpl.java
@@ -111,15 +111,15 @@ public class ReactiveRedisDataSourceImpl implements ReactiveRedisDataSource, Red
 
     @Override
     public Uni<TransactionResult> withTransaction(Function<ReactiveTransactionalRedisDataSource, Uni<Void>> tx,
-            String... keys) {
+            String... watchedKeys) {
         nonNull(tx, "tx");
-        notNullOrEmpty(keys, "keys");
-        doesNotContainNull(keys, "keys");
+        notNullOrEmpty(watchedKeys, "watchedKeys");
+        doesNotContainNull(watchedKeys, "watchedKeys");
         return redis.connect()
                 .onItem().transformToUni(connection -> {
                     ReactiveRedisDataSourceImpl singleConnectionDS = new ReactiveRedisDataSourceImpl(vertx, redis, connection);
                     TransactionHolder th = new TransactionHolder();
-                    return watch(connection, keys) // WATCH keys
+                    return watch(connection, watchedKeys) // WATCH keys
                             .chain(() -> connection.send(Request.cmd(Command.MULTI))
                                     .chain(x -> tx
                                             .apply(new ReactiveTransactionalRedisDataSourceImpl(singleConnectionDS, th)))
@@ -159,10 +159,10 @@ public class ReactiveRedisDataSourceImpl implements ReactiveRedisDataSource, Red
     @Override
     public <I> Uni<OptimisticLockingTransactionResult<I>> withTransaction(Function<ReactiveRedisDataSource, Uni<I>> preTx,
             BiFunction<I, ReactiveTransactionalRedisDataSource, Uni<Void>> tx, String... watchedKeys) {
+        nonNull(preTx, "preTx");
         nonNull(tx, "tx");
         notNullOrEmpty(watchedKeys, "watchedKeys");
         doesNotContainNull(watchedKeys, "watchedKeys");
-        nonNull(preTx, "preTx");
 
         return redis.connect()
                 .onItem().transformToUni(connection -> {

--- a/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/TransactionHolder.java
+++ b/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/TransactionHolder.java
@@ -39,11 +39,12 @@ public class TransactionHolder {
         List<Object> results = new ArrayList<>();
         for (int i = 0; i < mappers.size(); i++) {
             Response responsePart = response.get(i);
-            if (responsePart.type() == ResponseType.ERROR) {
+            if (responsePart == null || responsePart.type() != ResponseType.ERROR) {
+                // `null` is a valid result
+                results.add(mappers.get(i).apply(responsePart));
+            } else {
                 hasErrors = true;
                 results.add(responsePart.getDelegate()); // it's also an exception
-            } else {
-                results.add(mappers.get(i).apply(responsePart));
             }
         }
         return new OptimisticLockingTransactionResultImpl<>(discarded, hasErrors, input, results);

--- a/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/Validation.java
+++ b/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/Validation.java
@@ -127,13 +127,13 @@ public class Validation {
 
     public static void positiveOrZero(double amount, String name) {
         if (amount < 0) {
-            throw new IllegalArgumentException(String.format("`%s` must be greater or equal to zero`", name));
+            throw new IllegalArgumentException(String.format("`%s` must be greater or equal to zero", name));
         }
     }
 
     public static void isBit(int b, String name) {
         if (b != 0 && b != 1) {
-            throw new IllegalArgumentException(String.format("%s` must be either `0` or `1`", name));
+            throw new IllegalArgumentException(String.format("`%s` must be either `0` or `1`", name));
         }
     }
 

--- a/extensions/redis-client/runtime/src/test/java/io/quarkus/redis/datasource/OptimisticLockingTest.java
+++ b/extensions/redis-client/runtime/src/test/java/io/quarkus/redis/datasource/OptimisticLockingTest.java
@@ -347,4 +347,27 @@ public class OptimisticLockingTest extends DatasourceTestBase {
         assertThat(res.getPreTransactionResult()).isEqualTo("a");
     }
 
+    @Test
+    public void nullInResult() {
+        OptimisticLockingTransactionResult<Object> result = blocking.withTransaction(
+                ds -> null,
+                (ignored, tx) -> tx.value(String.class).get(key),
+                key);
+
+        assertThat(result.size()).isEqualTo(1);
+        assertThat(result.discarded()).isFalse();
+        assertThat((Object) result.get(0)).isNull();
+    }
+
+    @Test
+    public void nullInResultReactive() {
+        OptimisticLockingTransactionResult<Void> result = reactive.withTransaction(
+                ds -> Uni.createFrom().voidItem(),
+                (ignored, tx) -> tx.value(String.class).get(key),
+                key).await().indefinitely();
+
+        assertThat(result.size()).isEqualTo(1);
+        assertThat(result.discarded()).isFalse();
+        assertThat((Object) result.get(0)).isNull();
+    }
 }

--- a/extensions/redis-client/runtime/src/test/java/io/quarkus/redis/datasource/TransactionalValueCommandsTest.java
+++ b/extensions/redis-client/runtime/src/test/java/io/quarkus/redis/datasource/TransactionalValueCommandsTest.java
@@ -42,14 +42,16 @@ public class TransactionalValueCommandsTest extends DatasourceTestBase {
             string.append(key, "-1");
             string.get(key);
             string.strlen("k2");
+            string.get("nonexisting_key");
         });
-        assertThat(result.size()).isEqualTo(5);
+        assertThat(result.size()).isEqualTo(6);
         assertThat(result.discarded()).isFalse();
         assertThat(result.<Void> get(0)).isNull();
         assertThat((boolean) result.get(1)).isTrue();
         assertThat((long) result.get(2)).isEqualTo(7L);
         assertThat((String) result.get(3)).isEqualTo("hello-1");
         assertThat((long) result.get(4)).isEqualTo(7L);
+        assertThat((Object) result.get(5)).isNull();
     }
 
     @Test
@@ -60,15 +62,17 @@ public class TransactionalValueCommandsTest extends DatasourceTestBase {
                     .chain(() -> string.setnx("k2", "bonjour"))
                     .chain(() -> string.append(key, "-1"))
                     .chain(() -> string.get(key))
-                    .chain(() -> string.strlen("k2"));
+                    .chain(() -> string.strlen("k2"))
+                    .chain(() -> string.get("nonexisting_key"));
         }).await().atMost(Duration.ofSeconds(5));
-        assertThat(result.size()).isEqualTo(5);
+        assertThat(result.size()).isEqualTo(6);
         assertThat(result.discarded()).isFalse();
         assertThat(result.<Void> get(0)).isNull();
         assertThat((boolean) result.get(1)).isTrue();
         assertThat((long) result.get(2)).isEqualTo(7L);
         assertThat((String) result.get(3)).isEqualTo("hello-1");
         assertThat((long) result.get(4)).isEqualTo(7L);
+        assertThat((Object) result.get(5)).isNull();
     }
 
 }

